tidy3d/components/data/sim_data.py

```diff
@@ -978,9 +982,158 @@ def make_adjoint_sim(
+            unique_sources = {src.json(exclude={"source_time"}) for src in adj_srcs}
+            if len(unique_sources) > 1:
+                log.info(
+                    f"One monitor with {len(unique_sources)} adjoint sources. "
+                    "But, because of different source characteristics, "
+                    f"the monitor still requires f{len(unique_sources)} different adjoint sources."
+                    " No special handling."
+                )
+                return self._process_adjoint_sources_same_freq(sources_adj_dict)
+
+            # many sources differ only by source can be handled with broadband adjoint
+            return self._process_adjoint_sources_broadband(sources_adj_dict)
+
+        # typical case: several monitors with adjoint sources at same freq
+        return self._process_adjoint_sources_same_freq(sources_adj_dict)
+
+    def _process_adjoint_sources_same_freq(
+        self, sources_adj_dict: dict[str, list[Source]]
+    ) -> tuple[list[Source], xr.DataArray]:
+        """Process adjoint sources for the case of one adjoint source at several freqs."""
+
+        # map of monitor name to set of unique frequencies in the corresponding data
+        sources_unique_freqs = {
+            key: {src.source_time.freq0 for src in val} for key, val in sources_adj_dict.items()
+        }
+
+        def error_msg_pre(mnt_name: str) -> str:
+            """First part of error message for specific monitor."""
+            return f"Can't compute adjoint source for data from monitor '{mnt_name}'. "
+
+        error_msg_post = (
+            "Can only compute adjoint source for several monitor output if each"
+            "data has one frequency only."
+        )
+
+        # perform validation of the adjoint sources
+
+        # first, make sure each monitor data only has one frequency
+        for mnt_name, freqs in sources_unique_freqs.items():
+            if len(freqs) > 1:
+                raise ValueError(
+                    error_msg_pre(mnt_name)
+                    + f"Monitor has {len(freqs)} frequencies. "
+                    + error_msg_post
+                )
+
+        # then, more generally, ensure that all monitor data have the same frequency
+        all_freqs = set()
+        for mnt_name, freqs in sources_unique_freqs.items():
+            freq = tuple(freqs)[0]
+            all_freqs.add(freq)
+            if len(all_freqs) > 1:
+                raise ValueError(
+                    error_msg_pre(mnt_name)
+                    + error_msg_post
+                    + " Detected different frequencies in different monitors."
+                )
+
+        log.info("Several adjoint sources from different monitors, all with same single frequency.")
+
+        # passed validation, put all sources into a single list and set None for post-normalize amps
+        adj_srcs = []
+        for srcs in sources_adj_dict.values():
+            adj_srcs += srcs
```

Comment:
I would write `adj_srcs = sum(sources_adj_dict.values(), [])`.
Not sure if it's maybe less readable.

Reply:
learn something new every day :D. let's try this.